DCC-5632 Use Export from ID service as local cache

grant-guo opened the pull request:
@andricDu

Commit:
id class unit test

Commit:
unit test for ExportStringParser

dcc-jenkins commented:
Ready for review. 	- Type "ok to test" to initiate a PR build. 	- Type "test this please" to retest. 	- Type "add to whitelist" to always build on your PRs.	See https://wiki.jenkins-ci.org/display/JENKINS/GitHub+pull+request+builder+plugin

andricDu commented:
add to whitelist

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/model/FileID.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.Data;
+
+/**
+ * Created by gguo on 6/8/17.
```
you can remove these comments.

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/model/FileID.java:
```diff
@@ -0,0 +1,13 @@
+package org.icgc.dcc.release.job.id.model;
```
Your new files seem to be missing the license header.

Here is a copyright template you can use in IntelliJ:
```
Copyright (c) $today.year The Ontario Institute for Cancer Research. All rights reserved.

This program and the accompanying materials are made available under the terms of the GNU Public License v3.0.
You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

```

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/model/DonorID.java:
```diff
@@ -0,0 +1,17 @@
+ */
+@Data
+public class DonorID implements Serializable{
+    @NotNull
```
We prefer lombok's `@NonNull`

andricDu replied:
also one newline between class signature and first field/constant. 

You can see our general guidelines here: https://wiki.oicr.on.ca/display/DCCSOFT/Development+-+Conventions

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/model/DonorID.java:
```diff
@@ -0,0 +1,17 @@
+    @NotNull
+    private final String id;
+    @NotNull
+    private final String project;
```
Can you make sure you've set your IDE to use 2 spaces for indentation rather than 4.

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/model/MutationID.java:
```diff
@@ -0,0 +1,32 @@
+@Data
+public class MutationID implements Serializable {
+
+    @NotNull
```
`@NonNull` s for all of these guys

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/model/SampleID.java:
```diff
@@ -0,0 +1,18 @@
+ */
+@Data
+public class SampleID implements Serializable {
+    @NotNull
```
newline and `@NonNull`

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/model/SampleID.java:
```diff
@@ -0,0 +1,18 @@
+import java.io.Serializable;
+
+/**
+ * Created by gguo on 6/8/17.
```
delete this block comment

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/model/SpecimenID.java:
```diff
@@ -0,0 +1,18 @@
+import java.io.Serializable;
+
+/**
+ * Created by gguo on 6/8/17.
```
delete

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/model/SpecimenID.java:
```diff
@@ -0,0 +1,18 @@
+ */
+@Data
+public class SpecimenID implements Serializable {
+    @NotNull
```
`\n` and `@NonNull`

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/parser/ExportStringParser.java:
```diff
@@ -0,0 +1,27 @@
+
+    public Map<IDTYPE, String> parse(String str, GenerateIDTypeInstance<IDTYPE> generator){
+        return
+            Separators.getCorrespondingSplitter(Separators.NEWLINE).splitToList(str).stream().map(
```
Something like this would be much easier to read.
```java
import static org.icgc.dcc.common.core.util.Splitters.NEWLINE;
import static org.icgc.dcc.common.core.util.Splitters.TAB;
import static org.icgc.dcc.common.core.util.stream.Collectors.toImmutableMap;
.
.
.

  public Map<IDTYPE, String> parse(String str, GenerateIDTypeInstance<IDTYPE> generator) {
    return NEWLINE.splitToList(str).stream()
        .map(TAB::splitToList)
        .map(generator::generate)
        .collect(toImmutableMap(Entry::getKey, Entry::getValue));
  }

  public interface GenerateIDTypeInstance<IDTYPE> {
    Entry<IDTYPE, String> generate(List<String> fields);
  }
```

You can probably get away with using `Entry` and not have to bring in any apache commons. 
We also have a useful `dcc-common` project where we have a ton of useful things to use like splitters and collectors.

andricDu reviewed (changes requested):
Jenkins build is failing because it cannot find the dcc-id module that has your changes. You will need to open up a PR for dcc-id and then modify the current dcc-release[-job-id] pom to point to the pr-snapshot build of dcc-id. 

I've left some feedback regarding code style, lombok stuff, among other things. Once the dcc-id id stuff is references correctly and the project builds, I'll take a second look.

Commit:
code review comments

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/model/MutationID.java:
```diff
@@ -1,32 +1,46 @@
 
 /**
- * Created by gguo on 6/8/17.
+ Copyright (c) $today.year The Ontario Institute for Cancer Research. All rights reserved.
```
In IntelliJ go to `Editor` -> `Copyright` -> `Copyright Profiles` and use the template there.

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/parser/ExportStringParser.java:
```diff
@@ -1,27 +1,41 @@
 
 /**
- * Created by gguo on 6/8/17.
+ Copyright (c) $today.year The Ontario Institute for Cancer Research. All rights reserved.
```
copyright should be at tope of file.

Commit:
create ***Id object, then pass into spark map function

Commit:
mutation task stops using broadcast

Commit:
use spark dataframe for mutation data, use broadcast for donor, specimen and sample data

Commit:
log exceptions

Commit:
run testsuite instead of individual tests when jenkins is building

Commit:
add id prefix

Commit:
mutation task unit test and parallelize the broadcast preparation

Commit:
fixed a bug about putting the prefix to id string

andricDu commented:
@grant-guo can you investigate the failing unit test in jenkins

andricDu commented:
test this please

Commit:
merge dcc/develop and solve the conflict in DefaultJobContext

andricDu commented:
test this please

Commit:
don't write assembly_version to json string

andricDu commented on dcc-release-job/dcc-release-job-id/src/main/java/org/icgc/dcc/release/job/id/task/AddSurrogateMutationIdTask.java:
```diff
@@ -1,41 +1,156 @@
+                  mutation,
+                  mutationType,
+                  assemblyVersion,
+                  localIdClientFactory.create().createMutationId(
```
leaving a reminder here to change this so a client isn't instantiated for each missing mutationId.

Commit:
add missing json fields

Commit:
don't remove MutationID-related fields from ObjectNode in MutatationEntity class

grant-guo commented:
test this please

Commit:
1. add postgresql access configuration to Fathmm job
2. create a FathmmPredictor instance including FathmmRepository per executor
3. change how FathmmRepository creates Query objects in concurrent environment

grant-guo commented:
test this please

grant-guo commented:
test this please

Commit:
use synchronizers

Commit:
download reference file once per executor

Commit:
check if versionDir exists to avoid recreation

Commit:
download reference file no matter if verion dir exists

grant-guo commented:
test this please

grant-guo commented:
test this please

Commit:
call foreachPartition on RDD instead of mapPartitions then count

Commit:
clean comments

grant-guo commented:
test this please

grant-guo commented:
test this please

grant-guo commented:
test this please

grant-guo commented:
test this please

Commit:
repartition mutation ID rdd to 2

Commit:
synchronized on FathmmPredictor class instance instead of Object class instance
